### ISSUE TITLE
Connection lost gets receptor network on a not working state

### ISSUE THREAD

**Issue by elyezer** [project-receptor/python-receptor#79]
Considering a network like `controller -> node-a -> node-b -> node-c` and then running ping every second using controller as the ping peer node to ping node-c. Then stop either node-b or node-a and then after a while bringing it back on, the network will stop working and ping won't just come back to work. Also the following stack trace will be shown. To get the network to work again all the nodes needs to be restarted.

```
ERROR 2019-12-16 13:32:40,275 node-b base watch_queue: error received trying to write
Traceback (most recent call last):
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 39, in watch_queue
    await conn.send(msg)
  File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 31, in send
    await self.writer.drain()
  File "/usr/lib64/python3.6/asyncio/streams.py", line 339, in drain
    yield from self._protocol._drain_helper()
  File "/usr/lib64/python3.6/asyncio/streams.py", line 210, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost
Task exception was never retrieved
future: <Task finished coro=<serve() done, defined at /home/elyezer/code/receptor/receptor/receptor/connection/sock.py:51> exception=TypeError("object NoneType can't be used in 'await' expression",)>
Traceback (most recent call last):
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 39, in watch_queue
    await conn.send(msg)
  File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 31, in send
    await self.writer.drain()
  File "/usr/lib64/python3.6/asyncio/streams.py", line 339, in drain
    yield from self._protocol._drain_helper()
  File "/usr/lib64/python3.6/asyncio/streams.py", line 210, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/elyezer/code/receptor/receptor/receptor/connection/sock.py", line 53, in serve
    await factory().server(t)
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 128, in server
    await self.start_processing()
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 103, in start_processing
    return await self.write_task
  File "/home/elyezer/code/receptor/receptor/receptor/connection/base.py", line 43, in watch_queue
    return await conn.close()
TypeError: object NoneType can't be used in 'await' expression
```

**Comment by elyezer** [comment]
Steps to reproduce:

1. Open 5 terminal windows and run the following commands in order on each of the windows:

```
receptor \
    --node-id=controller \
    -d /tmp/controller \
    controller \
    --listen=receptor://127.0.0.1:9999
```

```
receptor \
    --node-id=node-a \
    -d /tmp/node-a \
    node \
    --listen=receptor://127.0.0.1:9998 \
    --peer=receptor://localhost:9999
```

```
receptor \
    --node-id=node-b \
    -d /tmp/node-b \
    node \
    --listen=receptor://127.0.0.1:9997 \
    --peer=receptor://localhost:9998
```

```
receptor \
    --node-id=node-c \
    -d /tmp/node-c \
    node \
    --listen=receptor://127.0.0.1:9996 \
    --peer=receptor://localhost:9997
```

```
receptor \
    -d /tmp/ping ping \
    --peer=receptor://127.0.0.1:9999 \
    --delay=1 \
    node-c
```

2. Then go to the second or third terminal and press Ctrl-c and wait some time. The ping running on the 5th terminal window will stop getting responses.

3. Restart the process that was finished on step 2. And observe that the ping won't continue unless all the receptor nodes and controller are restarted.

**Comment by elyezer** [comment]
I've rebased after the latest changes and I continue to observe the issue. I can't get the ping to work again until restart all the receptor node processes.

Another thing that I've observed is that when a node is brought back online it does not recognize the ping directive. I got the following debug logs by running the scenario described on previous comment then bringing `node-b` down and after couple seconds back online with the `--debug` option specified:

```
INFO 2019-12-17 14:46:29,933 node-b entrypoints Running as Receptor node with ID: node-b
INFO 2019-12-17 14:46:29,933 node-b controller Serving on receptor://127.0.0.1:9997
INFO 2019-12-17 14:46:29,934 node-b controller Connecting to peer receptor://localhost:9998
DEBUG 2019-12-17 14:46:29,937 node-b base sending HI
DEBUG 2019-12-17 14:46:29,938 node-b base waiting for HI
DEBUG 2019-12-17 14:46:29,940 node-b base sending routes
DEBUG 2019-12-17 14:46:29,940 node-b receptor Emitting Route Advertisements, excluding set()
DEBUG 2019-12-17 14:46:29,942 node-b base starting normal loop
DEBUG 2019-12-17 14:46:29,942 node-b receptor spawning message_handler
DEBUG 2019-12-17 14:46:29,946 node-b receptor message_handler: FramedMessage(msg_id=338220355935050702824898172092851698922, header={'sender': 'ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'recipient': 'node-c', 'route_list': ['ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'controller', 'node-a']}, payload=bytearray(b'{"message_id": "c0ed3724-6d7b-4ad5-a5b3-ad3dc0e82de0", "sender": "ca8ad9ef-0932-441d-b9c6-8b800b1841f8", "recipient": "node-c", "message_type": "directive", "timestamp": "2019-12-17T19:46:27.029037", "raw_payload": "2019-12-17T19:46:27.029037", "directive": "receptor:ping", "in_response_to": null, "ttl": 15, "serial": 1, "code": 0}'))
INFO 2019-12-17 14:46:29,947 node-b directive Received ping from ca8ad9ef-0932-441d-b9c6-8b800b1841f8
ERROR 2019-12-17 14:46:29,948 node-b receptor error in handle_message: Invalid directive -> 'receptor:ping'. Sending failure response back.
```

Without the `--debug` option it just prints:

```
ERROR 2019-12-17 14:48:45,064 node-b receptor error in handle_message: Invalid directive -> 'receptor:ping'. Sending failure response back.
```

**Comment by elyezer** [comment]
Here is a GIF version of what I am seeing, after restating node-a the ping flow never continue.

![receptor-ping-issue](https://user-images.githubusercontent.com/48132/71117815-ddeef680-21a4-11ea-880c-239f284282f1.gif)

**Comment by elyezer** [comment]
Today I gave this another try after the recent changes and I am still facing the issue.

It seems that either the node that is sending the message does not update the connection after the other node connection restarts or maybe it can be an error message not being properly handled:

```console
$ receptor --debug --node-id=node-b -d 
/tmp/node-b node --listen=receptor://127.0.0.1:9997 --peer=receptor://localhost:9998
INFO 2020-01-27 15:35:19,775 node-b entrypoints Running as Receptor node with ID: node-b
INFO 2020-01-27 15:35:19,775 node-b controller Serving on receptor://127.0.0.1:9997   
INFO 2020-01-27 15:35:19,776 node-b controller Connecting to peer receptor://localhost:9998
DEBUG 2020-01-27 15:35:19,780 node-b base sending HI                                  
DEBUG 2020-01-27 15:35:19,780 node-b base waiting for HI                              
DEBUG 2020-01-27 15:35:19,782 node-b base sending routes                              
DEBUG 2020-01-27 15:35:19,782 node-b receptor Emitting Route Advertisements, excluding set()
DEBUG 2020-01-27 15:35:19,784 node-b base starting normal loop                        
DEBUG 2020-01-27 15:35:19,784 node-b receptor spawning message_handler                
DEBUG 2020-01-27 15:35:19,788 node-b receptor message_handler: FramedMessage(msg_id=119436487780921991910312687304622836021, header={'sender': 'ca8ad9ef-0932-441d-b9c6-8b800
b1841f8', 'recipient': 'node-c', 'route_list': ['ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'controller', 'node-a']}, payload=bytearray(b
'{"message_id": "3f918961-6101-4a2b-9801-69db1e78ea5b", "sender": "ca8ad9ef-0932-441d-b9c6-8b800b1841f8", "recipient": "node-c", "message_type": "directive", "timestamp": "2
020-01-27T20:35:09.801732", "raw_payload": "2020-01-27T20:35:09.801732", "directive": "receptor:ping", "in_response_to": null, "ttl": 15, "serial": 1, "code": 0}'))
INFO 2020-01-27 15:35:19,790 node-b directive Received ping from ca8ad9ef-0932-441d-b9c6-8b800b1841f8
ERROR 2020-01-27 15:35:19,790 node-b receptor error in handle_message: Invalid directive -> 'receptor:ping'. Sending failure response back.

---
For some reason receptor fails to recognize the ping directive and then it seems to restart the loop.
I am adding these lines to improve readability for these logs
---

DEBUG 2020-01-27 15:35:21,510 node-b base waiting for HI                              
DEBUG 2020-01-27 15:35:21,513 node-b base sending HI                                  
DEBUG 2020-01-27 15:35:21,515 node-b base sending routes                              
DEBUG 2020-01-27 15:35:21,515 node-b receptor Emitting Route Advertisements, excluding set()
DEBUG 2020-01-27 15:35:21,529 node-b base starting normal loop                        
DEBUG 2020-01-27 15:35:21,531 node-b receptor spawning message_handler                
DEBUG 2020-01-27 15:35:21,531 node-b receptor message_handler: FramedMessage(msg_id=293241058218921198706150964326106198210, header={'cmd': 'ROUTE', 'id': 'node-c', 'capabil
ities': [['receptor_sleep', '1.0.0']], 'groups': [], 'edges': [['controller', 'node-a', 1], ['ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'controller', 1], ['node-b', 'node-c', 1
00], ['node-a', 'node-b', 1]], 'seen': ['node-b', 'node-c']}, payload=None)           
DEBUG 2020-01-27 15:35:21,532 node-b receptor Emitting Route Advertisements, excluding {'node-c', 'node-b'}
DEBUG 2020-01-27 15:35:21,545 node-b receptor message_handler: FramedMessage(msg_id=20092987850531595558000226240900103120, header={'cmd': 'ROUTE', 'id': 'node-c', 'capabili
ties': [['receptor_sleep', '1.0.0']], 'groups': [], 'edges': [['controller', 'node-a', 1], ['node-b', 'node-c', 1], ['ca8ad9ef-0932-441d-b9c6-8b800b1841f8', 'controller', 1]
, ['node-a', 'node-b', 1]], 'seen': ['node-b', 'node-c']}, payload=None)              
DEBUG 2020-01-27 15:35:21,545 node-b receptor Emitting Route Advertisements, excluding {'node-c', 'node-b'}
```

After the following line:

```
ERROR 2020-01-27 15:35:19,790 node-b receptor error in handle_message: Invalid directive -> 'receptor:ping'. Sending failure response back.
```

The process seems to be restarted and the HI and the loop seem to be restarted.

**Comment by elyezer** [comment]
I tried to reproduce this on latest devel (2a55ce14637927e837639a7139af2de62f745208) and I could not reproduce this anymore. I tried to randomly exit nodes and bring them back after waiting a random amount of time and the network was functional.

One concern is that the pending messages were "dumped" whenever the connection was established, this will cause a short burst on the communication. I will try to explore that more and will open a separated issue if needed.
